src/lib.rs

```diff
@@ -1,14 +1,207 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+use std::{cmp::Ordering, hash};
+
+pub use proc_macros::grouped_ordering;
+
+pub trait GroupedOrdering {
+    type Group: hash::Hash + Eq;
+
+    fn compare(&self, a: &Self::Group, b: &Self::Group) -> Ordering;
+}
+
+pub trait GroupedOrderable<TGroupedOrdering: GroupedOrdering> {
+    fn map_to_grouped_ordering(&self) -> TGroupedOrdering::Group;
```

Comment:
Then this is a trait for the "things that are orderable by a given grouped-ordering type" to implement (so declares sort of a structured relationship between the orderable things/grouped-ordering-specification type)

I didn't add any tests for this but part of the idea was that this should be a generic trait so that a given "orderable thing" type can implement this trait for multiple different grouped-ordering-specification types (ie "I (eg `String`) can be group-ordered according to not only different instances of a given grouped-ordering but also according to multiple different grouped-orderings")